torchtune/models/llama3/_tokenizer.py

```diff
@@ -0,0 +1,175 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Dict, List, Optional, Tuple
+
+from torchtune.data import Message, truncate
+from torchtune.modules.tokenizers import ModelTokenizer, TikTokenBaseTokenizer
+
+
+CL100K_PATTERN = r"""(?i:'s|'t|'re|'ve|'m|'ll|'d)|[^\r\n\p{L}\p{N}]?\p{L}+|\p{N}{1,3}| ?[^\s\p{L}\p{N}]+[\r\n]*|\s*[\r\n]+|\s+(?!\S)|\s+"""  # noqa
+
+PAD_ID = 0
+
+LLAMA3_SPECIAL_TOKENS = {
+    "<|begin_of_text|>": 128000,
+    "<|end_of_text|>": 128001,
+    "<|start_header_id|>": 128006,
+    "<|end_header_id|>": 128007,
+    "<|eot_id|>": 128009,
+    "<|eom_id|>": 128008,
+    "<|python_tag|>": 128255,
+}
+
+
+class Llama3Tokenizer(ModelTokenizer):
+    """
+    tiktoken tokenizer configured with Llama3 Instruct's special tokens, as described in
+    https://llama.meta.com/docs/model-cards-and-prompt-formats/meta-llama-3
+
+    Args:
+        path (str): Path to pretrained tiktoken tokenizer file.
+        special_tokens (Optional[Dict[str, int]]): mapping containing special text tokens and
+            their registered token IDs. If left as None, this will be set to the canonical
+            Llama3 special tokens.
+    """
+
+    def __init__(
+        self,
+        path: str,
+        special_tokens: Optional[Dict[str, int]] = None,
+    ):
+        self.special_tokens = special_tokens or LLAMA3_SPECIAL_TOKENS
+
+        # Encode BOS and EOS, define pad ID
+        self.bos_id = self.special_tokens["<|begin_of_text|>"]
+        self.eos_id = self.special_tokens["<|end_of_text|>"]
+        self.pad_id = PAD_ID
+
+        # Encode extra special tokens
+        self.start_header_id = self.special_tokens["<|start_header_id|>"]
+        self.end_header_id = self.special_tokens["<|end_header_id|>"]
+        self.eot_id = self.special_tokens["<|eot_id|>"]
```

Comment:
Not a huge deal, but I feel like this isn't the most configurable way to do this. Like we claim to support custom special tokens, but they also need to be a superset of LLAMA3_SPECIAL_TOKENS, which is not immediately obvious and requires some digging around to figure out. I guess the obvious alternative here is something like `extra_special_tokens` which tbh I don't love either. Maybe just add an assert or something with an explicit error message about which keys must be present when using this tokenizer.

Reply:
Yeah, I think in the ruckus your original assert checks got thrown out. let me add that back in